src/utils/chartJs/CustomTooltip.jsx

```diff
@@ -1,57 +1,85 @@
+  '&::before': { bottom: '100%', borderBottomColor: 'var(--bg-color)' },
+  '&.xright': {
+    '--trans-x': 'calc(-100% + 2 * var(--tip-size))',
+    '&::before': { right: 'var(--tip-size)' },
+  },
+  '&.xcenter': {
+    '--trans-x': '-50%',
+    '&::before': { left: 'calc(50% - var(--tip-size))' },
+  },
+  '&.xleft': {
+    '--trans-x': 'calc(-2 * var(--tip-size))',
+    '&::before': { left: 'var(--tip-size)' },
+  },
+};
 const styles = {
   tooltip: {
+    '--bg-color': 'rgba(0, 0, 0, .8)',
+    '--tip-size': '6px',
     position: 'absolute',
     padding: '6px',
-    background: 'rgba(0, 0, 0, .8)',
+    background: 'var(--bg-color)',
     color: 'white',
     borderRadius: '4px',
     pointerEvents: 'none',
+    transform: 'translate(var(--trans-x), var(--trans-y))',
+    '& a': {
+      pointerEvents: 'auto',
+    },
+    '&::before': {
+      position: 'absolute',
+      content: '""',
+      borderWidth: 'var(--tip-size)',
+      borderStyle: 'solid',
+      borderColor: 'transparent',
+    },
+    '&.ytop': {
+      ...topAligned,
+    },
+    '&.ycenter': {
+      '--trans-y': '-50%',
+      '&::before': { top: 'calc(50% - var(--tip-size))' },
+      '&.xright': {
+        '--trans-x': 'calc(-1 * (100% + var(--tip-size)))',
+        '&::before': { left: '100%', borderLeftColor: 'var(--bg-color)' },
+      },
+      '&.xleft': {
+        '--trans-x': 'var(--tip-size)',
+        '&::before': { right: '100%', borderRightColor: 'var(--bg-color)' },
+      },
+    },
+    '&.ybottom': {
+      ...topAligned,
+      '--trans-y': 'calc(-100% - var(--tip-size))',
+      '&::before': { top: '100%', borderTopColor: 'var(--bg-color)' },
+    },
   },
   tooltipKey: {
     display: 'inline-block',
     width: '10px',
     height: '10px',
     marginRight: '10px',
   },
+  lockMessage: {
+    color: '#ccc',
+  },
 };
 
 class CustomTooltip extends React.Component {
-  constructor(props) {
-    super(props);
-    this.tooltip = React.createRef();
-    this.state = { width: 0, height: 0 };
-  }
-
-  componentDidMount() {
-    this.setState({
-      width: this.tooltip.current.clientWidth,
-      height: this.tooltip.current.clientHeight,
-    });
-  }
-
   render() {
-    const { classes, tooltipModel, series, canvas } = this.props;
+    const { classes, tooltipModel, series, canvas, isLocked } = this.props;
 
     if (tooltipModel.opacity === 0) return null;
 
-    let top = canvas.offsetTop + tooltipModel.caretY;
-    let left = canvas.offsetLeft + tooltipModel.caretX;
-
-    if (tooltipModel.yAlign === 'bottom') {
-      top -= this.state.height;
-    } else if (tooltipModel.yAlign === 'center') {
-      top -= 0.5 * this.state.height;
-    }
-
-    if (tooltipModel.xAlign === 'right') {
-      left -= this.state.width;
-    } else if (tooltipModel.xAlign === 'center') {
-      left -= 0.5 * this.state.width;
-    }
-
+    const top = canvas.offsetTop + tooltipModel.caretY;
+    const left = canvas.offsetLeft + tooltipModel.caretX;
```

Comment:
I love your challenges.
I've previously encountered [JSS Dynamic Values](https://cssinjs.org/react-jss?v=v10.0.0-alpha.16#dynamic-values) and after reading your comment, I tried to applied it to our code so we could get rid of not just calculations but all the inline style made inside the component. Unfortunately I wasn't lucky to make it work at the moment, not even with testing commands like `color: props => 'blue'`. I'll see what am I doing it wrong, and if I can't make it work in this PR, it will be in the next ones.

Reply:
@klahnakoski 
I finally made it work, *but* we need to install `@material-ui/styles`, and import `withStyles` from `@material-ui/styles` instead of `@material-ui/core/styles`.

As it's stablished [here](https://material-ui.com/css-in-js/basics/) (at the top in the warning note), `@material-ui/styles` is still an alpha version, so I'm not sure if we should add that dependency now. Let me know what should be best, please.

Sources: [original issue](https://github.com/mui-org/material-ui/issues/7633) and [merged PR](https://github.com/mui-org/material-ui/pull/13503).

Reply:
My mistake.  I assumed the`@material-ui/core/styles` was capable of accepting dynamic values.

* Can **we** simply set the HTML element `style` property without using classes, and without using any of these CSS tools?
* What is the compiled size difference when using `@material-ui/styles` vs `@material-ui/core/styles` , or any other dynamic CSS solution?

Reply:
save it for the next PR

Reply:
Got it,  I'll work on it for the next PR. But answering your questions, anyway:

> Can **we** simply set the HTML element `style` property without using classes, and without using any of these CSS tools?

Already thinking about this.

> What is the compiled size difference when using `@material-ui/styles` vs `@material-ui/core/styles` , or any other dynamic CSS solution?

`./build` (using `@material-ui/styles`and without it): **8.7M** vs **8.4M**